Code/Framework/AzNetworking/AzNetworking/UdpTransport/UdpHeartbeatThread.cpp

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <AzNetworking/UdpTransport/UdpHeartbeatThread.h>
+#include <AzNetworking/UdpTransport/UdpNetworkInterface.h>
+
+
+namespace AzNetworking
+{
+    // HeartbeatThreadUpdateRateMs is how often the timed thread's OnUpdate call will be executed.
+    // Heartbeats will also be sent this often while the mainthread is blocked (ex: loading a level)
+    static constexpr AZ::TimeMs HeartbeatThreadUpdateRateMs{ 500 };
+
+    UdpHeartbeatThread::UdpHeartbeatThread()
+        : TimedThread("UdpHeartbeatThread", HeartbeatThreadUpdateRateMs)
+    {
+    }
+
+    UdpHeartbeatThread::~UdpHeartbeatThread()
+    {
+        Stop();
+        Join();
+    }
+
+    void UdpHeartbeatThread::OnStart()
+    {
+        ;
+    }
+
+    void UdpHeartbeatThread::OnStop()
+    {
+        ;
+    }
+
+    void UdpHeartbeatThread::OnUpdate([[maybe_unused]]AZ::TimeMs updateRateMs)
+    {
+        AZ::TimeMs currentTime = AZ::GetElapsedTimeMs();
+
+        for (auto udpInterface : m_udpNetworkInterfaces)
+        {
+            AZ::TimeMs timeSinceSystemTick = currentTime - udpInterface->GetLastSystemTickUpdate();
+            if (timeSinceSystemTick > HeartbeatThreadUpdateRateMs)
+            {
+                // Main-thread is blocked; send a heartbeat from this thread.
+                auto sendNetworkUpdates = [](AzNetworking::IConnection& connection)
+                {
+                    if (connection.GetConnectionRole() == ConnectionRole::Connector)
+                    {
+                        connection.SendUnreliablePacket(CorePackets::HeartbeatPacket(false));
+                    }
+                };
+
+                udpInterface->GetConnectionSet().VisitConnections(sendNetworkUpdates);
+            }
+        }
```

Comment:
1. Seems like this should lock m_mutex. Right now, I don't think it's providing any protection against multi-threaded access to m_udpNetworkInterfaces.
 
2. Why is this sending a heartbeat for *each* interface that hasn't sent a heartbeat in a while? Is this because the different interfaces might be going to different servers? Some comments explaining the concept of this loop would be helpful here. Also, it's not clear to me, is the assumption that the connections might also be sending heartbeat packets from other threads (like the main thread) as well? If so, that should definitely get spelled out either here or at the top of the file since it's not obvious just from reading this code. Maybe that's what the comments about "sends a heartbeat if the main thread is blocked" is supposed to mean, but if so, I wasn't quite getting that meaning from that comment and this logic. It isn't really clear how it would *only* be sending it here if the main thread is blocked, or for that matter, *why* we wouldn't just send it here all the time.

Reply:
Aye, the name "heartbeat" is unfortunate, because these DO NOT happen at regular intervals like an **actual** heartbeat

The cases a heartbeat is sent is...
_Note: A request heartbeat is packet which when received, the receiver will sent back a heartbeat._

1. Server will send a non-request heartbeat if it has received 10 packets (_net_UdpMaxUnackedPacketCount_) without having sent one back in a while. This is to just acknowledge to the client "hey, you're sending me stuff, I'm still here, but just haven't had a need to send anything back to you in a while"
2. Client will sent a 5 request heartbeats  (_net_UdpUnackedHeartbeats_) over 10 seconds (_net_UdpDefaultTimeoutMs_)  if it's worried the server has died. "Hey server, I'm starting to get worried, you still there?"
3. NEW Connectors (usually clients) will send heartbeats to the server if their mainthread is blocked (example: level loading) and so for sure have not sent any packets in a while. 

This does make me want to fix the system further to simply move more of the logic to the heartbeat thread, instead of what feels more like a one-off patch.

Reply:
It's more of a "keep alive" than a heartbeat, but I can see that a consistent heartbeat would be wasteful in a typical game scenario where we're constantly receiving frequent network traffic from clients/servers.

Reply:
I agree we should lock `m_mutex` unless there's some kind of gaurantee no one will try to register/unregister during OnUpdate

Reply:
> It's more of a "keep alive" than a heartbeat, but I can see that a consistent heartbeat would be wasteful in a typical game scenario where we're constantly receiving frequent network traffic from clients/servers.

I'd like to change the packet from CorePackets::Heartbeat to CorePackets::KeepAlive. Conceptually, that's exactly correct.